dbms/src/Functions/FunctionsGrouping.h

```diff
@@ -0,0 +1,268 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Columns/ColumnNullable.h>
+#include <Common/typeid_cast.h>
+#include <Core/ColumnNumbers.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <common/types.h>
+#include <tipb/metadata.pb.h>
+#include <Interpreters/Context.h>
+
+#include <magic_enum.hpp>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int TOO_LESS_ARGUMENTS_FOR_FUNCTION;
+extern const int TOO_MANY_ARGUMENTS_FOR_FUNCTION;
+} // namespace ErrorCodes
+
+using ResultType = UInt8;
+
+class FunctionGrouping : public IFunctionBase
+    , public IExecutableFunction
+    , public std::enable_shared_from_this<FunctionGrouping>
+{
+public:
+    static constexpr auto name = "grouping";
+    using ArgType = UInt64; // arg type should always be UInt64
+
+    FunctionGrouping(const DataTypes & argument_types_, const DataTypePtr & return_type_, const tipb::Expr & expr)
+        : argument_types(argument_types_)
+        , return_type(return_type_)
+    {
+        tipb::GroupingFunctionMetadata meta;
+        if (!meta.ParseFromString(expr.val()))
+            throw Exception("Grouping function decodes meta data fail");
+
+        mode = static_cast<tipb::GroupingMode>(meta.mode());
+        size_t num = meta.grouping_marks_size();
+
+        if (num <= 0)
+            throw Exception("number of grouping_ids should be greater than 0");
+
+        if (mode == 1 || mode == 2)
+            meta_grouping_id = meta.grouping_marks()[0];
+        else
+        {
+            for (size_t i = 0; i < num; ++i)
+                meta_grouping_marks.insert(meta.grouping_marks()[i]);
+        }
+    }
+
+    String getName() const override { return name; }
+    const DataTypes & getArgumentTypes() const override { return argument_types; }
+    const DataTypePtr & getReturnType() const override { return return_type; }
+
+    ExecutableFunctionPtr prepare(const Block & /*sample_block*/) const override
+    {
+        return std::const_pointer_cast<FunctionGrouping>(shared_from_this());
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        NullPresence null_presence = getNullPresense(block, arguments);
```

Comment:
> Why can't useDefulatImplForNulls and useDefaultImplForConstant?

I think we can use `useDefaultImplForConstant`, however, we'd better to handle nullable column in the function so that we can check the validation that no nullable column should be received in `grouping` function.